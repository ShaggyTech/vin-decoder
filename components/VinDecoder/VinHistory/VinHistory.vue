<script lang="ts">
import { defineComponent } from '@vue/composition-api';
import { historySetup } from '@/compositions/useHistoryStore';
/* Components */
import VinResults from '@/components/VinDecoder/VinResults/VinResults.vue';

export default defineComponent({
  name: 'VinHistory',
  components: { VinResults },

  setup(_, { root: { $accessor } }) {
    const { history, clearHistory } = historySetup($accessor);

    return {
      history,
      clearHistory
    };
  }
});
</script>

<template>
  <v-card
    v-if="history.length > 0"
    class="history-card mx-auto my-4"
    max-width="600"
  >
    <v-card-title class="history-card__title headline">
      Search History
      <v-spacer />
      <v-btn
        v-if="history.length > 0"
        x-small
        class="btn__clear-history mx-2 my-2"
        color="deep-orange darken-4"
        @click.prevent="clearHistory"
      >
        Clear History
      </v-btn>
    </v-card-title>
    <v-expansion-panels accordian focusable hover flat tile>
      <v-expansion-panel v-for="(item, key) in history" :key="key">
        <v-expansion-panel-header class="title primary darken-4">
          {{ item.VIN }}
        </v-expansion-panel-header>
        <v-expansion-panel-content>
          <vin-results
            :raw-results.sync="item.results"
            color="primary darken-4"
            tile
          />
        </v-expansion-panel-content>
      </v-expansion-panel>
    </v-expansion-panels>
  </v-card>
</template>
