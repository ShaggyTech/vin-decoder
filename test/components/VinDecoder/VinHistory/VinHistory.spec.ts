import { mount } from '@vue/test-utils';
import VinHistory from '@/components/VinDecoder/VinHistory/VinHistory.vue';

const factory = (options: object) => {
  return mount(VinHistory, {
    ...options
  });
};

const getMocks = (history: Array<any> = []) => {
  return {
    $accessor: {
      history: {
        history,
        clearHistory() {
          return (this.history = []);
        }
      }
    }
  };
};

const mockHistory = [{ VIN: 'TESTVIN', results: { Make: 'VW' } }];

describe('VinHistory Component Tests', () => {
  test('component is invisible if history is empty', () => {
    const wrapper = factory({
      sync: false,
      mocks: { ...getMocks() }
    });

    expect(wrapper.vm.$el.textContent).toEqual('');
    expect(wrapper.element).toMatchSnapshot();
  });

  test('component is visible if history contains items', () => {
    const wrapper = factory({
      sync: false,
      mocks: { ...getMocks(mockHistory) }
    });

    expect(wrapper.find('.history-card').exists()).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  test('history can be cleared and component is invisible after clearing', async () => {
    const wrapper = factory({
      sync: false,
      mocks: { ...getMocks(mockHistory) }
    });

    // history exists
    expect(wrapper.vm.$accessor.history.history).toEqual(mockHistory);
    // component is visible
    expect(wrapper.find('.history-card').exists()).toBe(true);

    // Click the button triggering the clearHistory method
    wrapper.find('.btn__clear-history').trigger('click');
    await wrapper.vm.$nextTick();

    // history is cleared
    expect(wrapper.vm.$accessor.history.history).toEqual([]);
    // component is invisible after clearing
    expect(wrapper.vm.$el.textContent).toEqual('');

    expect(wrapper.element).toMatchSnapshot();
  });
});
