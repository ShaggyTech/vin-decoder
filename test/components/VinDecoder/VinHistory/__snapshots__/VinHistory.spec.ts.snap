// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VinHistory Component Tests component is invisible if history is empty 1`] = `<!---->`;

exports[`VinHistory Component Tests component is visible if history contains items 1`] = `
<div
  class="history-card mx-auto my-4 v-card v-sheet theme--light"
  style="max-width: 600px;"
>
  <div
    class="v-card__title history-card__title headline"
  >
    
    Search History
    
    <div
      class="spacer"
    />
     
    <button
      class="btn__clear-history mx-2 my-2 v-btn v-btn--contained theme--light v-size--x-small deep-orange darken-4"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        
      Clear History
    
      </span>
    </button>
  </div>
   
  <div
    accordian=""
    class="v-item-group theme--light v-expansion-panels v-expansion-panels--flat v-expansion-panels--hover v-expansion-panels--focusable v-expansion-panels--tile"
  >
    <div
      aria-expanded="false"
      class="v-expansion-panel"
    >
      <button
        class="v-expansion-panel-header title primary darken-4"
        type="button"
      >
        
        TESTVIN
      
        <transition-stub
          name="fade-transition"
        >
          <div
            class="v-expansion-panel-header__icon"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light"
            >
              $expand
            </i>
          </div>
        </transition-stub>
      </button>
       
      <transition-stub
        name="expand-transition"
      >
        <!---->
      </transition-stub>
    </div>
  </div>
</div>
`;

exports[`VinHistory Component Tests history can be cleared and component is invisible after clearing 1`] = `<!---->`;
